synapse/storage/databases/main/end_to_end_keys.py

```diff
@@ -114,6 +133,46 @@ async def get_e2e_device_keys_for_federation_query(
         """
         now_stream_id = self.get_device_stream_token()
 
+        # We need to be careful with the caching here, as we need to always
+        # return *all* persisted devices, however there may be a lag between a
+        # new device being persisted and the cache being invalidated.
+        cached_results = (
+            self._get_e2e_device_keys_for_federation_query_inner.cache.get_immediate(
+                user_id, None
+            )
+        )
+        if cached_results is not None:
+            # Check that there have been no new devices added by another worker
+            # after the cache. This should be quick as there should be few rows
+            # with a higher stream ordering.
+            sql = """
+                SELECT user_id FROM device_lists_stream
+                WHERE stream_id >= ? AND user_id = ?
+            """
+            rows = await self.db_pool.execute(
+                "get_e2e_device_keys_for_federation_query_check",
+                None,
+                sql,
+                now_stream_id,
+                user_id,
+            )
+            if not rows:
+                # No new rows, so cache is still valid.
+                return now_stream_id, cached_results
+
+            # There has, so let's invalidate the cache and run the query.
+            self._get_e2e_device_keys_for_federation_query_inner.invalidate((user_id,))
+
+        results = await self._get_e2e_device_keys_for_federation_query_inner(user_id)
+
+        return now_stream_id, results
```

Comment:
I'm having trouble figuring out whether this is correct.

`cached_results` will come from a time when the device stream token was less than `now_stream_id`. But we only discard the cached result if there are rows with a stream ordering >= `now_stream_id`, so wouldn't we be overlooking changes between `cached_results`'s stream id and `now_stream_id`?

Reply:
I think that case is fine, as for all changes less than `now_stream_id` we should have invalidated the cache?

Reply:
I see, if `now_stream_id` can't advance without invalidating the cache then it looks okay.
We could probably add a comment to that effect.

Reply:
Yes, it makes enough sense.